src/db.c

```diff
@@ -1516,31 +1507,50 @@ int keyIsExpired(redisDb *db, robj *key) {
  * is via lookupKey*() family of functions.
  *
  * The behavior of the function depends on the replication role of the
- * instance, because slave instances do not expire keys, they wait
- * for DELs from the master for consistency matters. However even
- * slaves will try to have a coherent return value for the function,
- * so that read commands executed in the slave side will be able to
- * behave like if the key is expired even if still present (because the
- * master has yet to propagate the DEL).
+ * instance, because replicas do not expire keys. They wait for DELs
+ * from the master for consistency matters. The exception is write
+ * operations to writable replicas. However, even read-only replicas
+ * will try to have a coherent return value of the function, so that
+ * read commands executed on the replica will be able to behave as if
+ * the key is expired even if still present (because the master has yet
+ * to propagate the DEL).
  *
  * In masters as a side effect of finding a key which is expired, such
  * key will be evicted from the database. Also this may trigger the
  * propagation of a DEL/UNLINK command in AOF / replication stream.
  *
  * The return value of the function is 0 if the key is still valid,
  * otherwise the function returns 1 if the key is expired. */
-int expireIfNeeded(redisDb *db, robj *key) {
+int expireIfNeeded(redisDb *db, robj *key, int expire_on_replica) {
+
+    /* Deleting expired keys on a replica makes the replica inconsistent with
+     * the master. The reason it's allowed for write commands is to make
+     * writable replicas behave consistently. It shall not be used in readonly
+     * commands. Modules are accepted so that we don't break old modules. */
+    if (expire_on_replica) {
```

Comment:
only one suggestion: remove the `expire_on_replica` argument, just check if this is a writable-replica in `expireIfNeeded` by itself, do not let the writable-replica issue pollute the `lookupKeyRead/Write` function.

Reply:
I have tried, but if we remove this parameter, we instead need an `is_write` parameter. Otherwise we can't easily fulfil the request to keep the expired key on lookup for read.

Reply:
I'll rename it to `force_delete_expired` and move the assert to `lookupKey()`. I agree, the meaning of the parameter is more clear that way.